arch/xtensa/src/esp32/esp32_aes.c

```diff
@@ -0,0 +1,1051 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <debug.h>
+#include <assert.h>
+#include <semaphore.h>
+
+#include <nuttx/semaphore.h>
+#include <nuttx/crypto/crypto.h>
+
+#include "xtensa.h"
+#include "hardware/esp32_aes.h"
+#include "hardware/esp32_dport.h"
+#include "esp32_aes.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define AES_BLK_SIZE                    (16)
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static bool g_aes_inited;
+static sem_t g_aes_sem = SEM_INITIALIZER(1);
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: aes_hw_setkey
+ *
+ * Description:
+ *   Set AES hardware key and encryption/decryption mode
+ *
+ * Input Parameters:
+ *   aes     - AES object data pointer
+ *   encrypt - True: encryption mode; False: decryption mode
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void aes_hw_setkey(struct esp32_aes_s *aes, bool encrypt)
+{
+  int i;
+  uint32_t cryptbits = encrypt ? 0 : AES_MODE_DECRYPT;
+  uint32_t keybits = (aes->keybits / 64) - 2;
+  uint32_t keywords = aes->keybits / 32;
+
+  putreg32(cryptbits | keybits, AES_MODE_REG);
+
+  for (i = 0; i < keywords; ++i)
+    {
+      putreg32(aes->key[i], AES_KEY_BASE + i * 4);
+    }
+}
+
+/****************************************************************************
+ * Name: aes_hw_cypher
+ *
+ * Description:
+ *   Process AES hardware encryption/decryption.
+ *
+ * Input Parameters:
+ *   s - Input data pointer
+ *   d - Output buffer pointer
+ *
+ * Returned Value:
+ *   None.
+ *
+ ****************************************************************************/
+
+static void aes_hw_cypher(const uint8_t *s, uint8_t *d)
+{
+  uint32_t buffer[AES_BLK_SIZE / 4];
+
+  memcpy(buffer, s, AES_BLK_SIZE);
+
+  putreg32(buffer[0], AES_TEXT_BASE + 0);
+  putreg32(buffer[1], AES_TEXT_BASE + 4);
+  putreg32(buffer[2], AES_TEXT_BASE + 8);
+  putreg32(buffer[3], AES_TEXT_BASE + 12);
+
+  putreg32(AES_START_OPT, AES_START_REG);
+
+  while (getreg32(AES_IDLE_REG) != AES_IDLE_STATE)
+    {
+      ;
```

Comment:
This ";" inside is not necessary. You can put it after the "}" if you want, but also not necessary!